### ISSUE TITLE
[BUG] HTMLReport throws error on _some_ tests

### ISSUE THREAD

**Issue by flunderpero** [microsoft/playwright#10732]
**Context:**
## System:
 - OS: Linux 5.10 Ubuntu 20.10 (Groovy Gorilla)
 - Memory: 1.24 GB / 3.84 GB
 - Container: Yes
## Binaries:
 - Node: 14.18.1 - /usr/bin/node
 - Yarn: 3.1.0 - /usr/bin/yarn
 - npm: 6.14.15 - /usr/bin/npm
## Languages:
 - Bash: 5.0.17 - /usr/bin/bash

**Describe the bug**

For about 20% of all tests the page is blank (after clicking on the test) find the following error in the console:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'getData')
    at P (playwright.html:2)
    at playwright.html:2
    at playwright.html:2
    at Mi (playwright.html:2)
    at n.unstable_runWithPriority (playwright.html:2)
    at Wo (playwright.html:2)
    at Ri (playwright.html:2)
    at playwright.html:2
    at Q (playwright.html:2)
    at MessagePort.w.port1.onmessage (playwright.html:2)
```

I don't see any pattern. [Edit] It happens on failed and on succeeded tests.

**Comment by dgozman** [comment]
Looks very similar to #10713. If you have a repro, that would be great!

**Comment by flunderpero** [comment]
It seems to affect tests with `test.step` only.

**Comment by mxschmitt** [comment]
Closing since we can't act on this without a repro. Please create a new issue if you have a repro, thanks!

**Comment by gwittel** [comment]
I don't have a full repro yet, but this looks very similar.  In a scenario where I have an assertion failure inside a `test.step()`, I get this error:

```
TypeError: Cannot read properties of undefined (reading 'toString')

   at node_modules/@playwright/test/lib/reporters/base.js:279

  277 |         } else {
  278 |           if (attachment.contentType.startsWith('text/')) {
> 279 |             let text = attachment.body.toString();
      |                                        ^
  280 |             if (text.length > 300) text = text.slice(0, 300) + '...';
  281 |             resultLines.push(_safe.default.cyan(`    ${text}`));
  282 |           }

    at formatFailure (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:279:40)
    at /Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:204:19
    at Array.forEach (<anonymous>)
    at ListReporter._printFailures (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:203:14)
    at ListReporter.epilogue (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:194:77)
    at ListReporter.onEnd (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/list.js:181:10)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Multiplexer.onEnd (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/multiplexer.js:77:7)
    at Runner._run (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/runner.js:435:7)
```

If I tweak the `base.js` to check that `attachment.body` is defined before building the attachment, it works.  I suspect from the calling side, its our script trying to attach an undefined attachment body.  As a caller I've added a guard, but it would be good to have it on the receiver side.  I see the underlying typescript has `attachment.body!.toString()`, but that compiles away.

**Comment by rwoll** [comment]
> I don't have a full repro yet, but this looks very similar.  In a scenario where I have an assertion failure inside a `test.step()`, I get this error:
> 
> 
> 
> ```
> 
> TypeError: Cannot read properties of undefined (reading 'toString')
> 
> 
> 
>    at node_modules/@playwright/test/lib/reporters/base.js:279
> 
> 
> 
>   277 |         } else {
> 
>   278 |           if (attachment.contentType.startsWith('text/')) {
> 
> > 279 |             let text = attachment.body.toString();
> 
>       |                                        ^
> 
>   280 |             if (text.length > 300) text = text.slice(0, 300) + '...';
> 
>   281 |             resultLines.push(_safe.default.cyan(`    ${text}`));
> 
>   282 |           }
> 
> 
> 
>     at formatFailure (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:279:40)
> 
>     at /Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:204:19
> 
>     at Array.forEach (<anonymous>)
> 
>     at ListReporter._printFailures (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:203:14)
> 
>     at ListReporter.epilogue (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/base.js:194:77)
> 
>     at ListReporter.onEnd (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/list.js:181:10)
> 
>     at processTicksAndRejections (node:internal/process/task_queues:96:5)
> 
>     at Multiplexer.onEnd (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/reporters/multiplexer.js:77:7)
> 
>     at Runner._run (/Users/gwittel/src/tw/repos/utils/testing/node_modules/@playwright/test/lib/runner.js:435:7)
> 
> ```
> 
> 
> 
> If I tweak the `base.js` to check that `attachment.body` is defined before building the attachment, it works.  I suspect from the calling side, its our script trying to attach an undefined attachment body.  As a caller I've added a guard, but it would be good to have it on the receiver side.  I see the underlying typescript has `attachment.body!.toString()`, but that compiles away.

@gwittel I think that particular case is fixed by https://github.com/microsoft/playwright/pull/11414 (undefined and empty string were being treated equivalently when they shouldn't have). You can try the patch via `npm i @playwright/test@next`

**Comment by rwoll** [comment]
NB: Passing undefined for both path and body should result in an error of the test case, but not a crash of the runner. 

Before the fix linked above, the problem was empty string was not triggering the more graceful error (since empty isn't undefined), and was instead getting transformed by internal code that broke some assumptions. On `@next`, undefined for both body and string should result in a graceful test-level error, and an empty string as the body should not trigger any error and will be attached as an inline attachment with an empty body.

**Comment by rwoll** [comment]
@gwittel While looking more carefully at the stack trace you linked, I found another case that's potentially what you were hitting (if it's not the #11414 case mentioned in my previous comment)…or something else! Both seem to produce the same stack trace 😄 

I've just filed it as: https://github.com/microsoft/playwright/issues/11565. It could always be something else, so a repro will be best if you can provide one!

Please see https://github.com/microsoft/playwright/issues/11565 and comment there with the following:

1. Playwright Test Version you are using (`npm ls @playwright/test`)
2. how the test is attaching artifacts. (Is it a manual push onto the attachments array, or is it using the new [v1.18 `async testInfo.attach(…)` method](https://playwright.dev/docs/api/class-testinfo#test-info-attach)) 
3. Are you trying to attach an inline attachment (i.e. `body` parameter) or a path based attachment (i.e. `path` parameter)
4. Repro if you narrow down to one.

In general, the newer [v1.18 `async testInfo.attach(…)` method](https://playwright.dev/docs/api/class-testinfo#test-info-attach) should fail the test gracefully and provide better error messages around a few different cases and assumptions the reporters make, so depending on what you're currently doing that might be a good option. (The new v1.18 API has a few fixes in the works, so you might want to try the `@next` release if it's not working for you on 1.18 and open issues if `@next` is still not working).

Thanks!

**Comment by gwittel** [comment]
> @gwittel I think that particular case is fixed by #11414 (empty string was getting turned into an undefined when it shouldn't have). You can install one of the nightly builds via `npm i @playwright/test@next` and re-run to see if that resolves the cases you were hitting.

Thank you so much for looking at this.  I don’t have access at the moment but it’s highly likely #11414 will fix it. In my case it was attaching an empty string body. For the moment I had added a guard to avoid attaching empty values.

**Comment by gwittel** [comment]
Works in the @next tag.  Will comment on #11565 with the details.
